hmy/blockchain.go

```diff
@@ -73,6 +73,51 @@ func (hmy *Harmony) GetBlockSigners(
 	return committee.Slots, mask, nil
 }
 
+// DetailedBlockSignerInfo contains all of the block singing information
+type DetailedBlockSignerInfo struct {
+	// Signers is a map of addresses in the Signers for the block to
+	// all of the serialized BLS keys that signed said block.
+	Signers map[common.Address][]bls.SerializedPublicKey
+	// Committee when the block was signed.
+	Committee shard.SlotList
+	// TotalKeysSigned is the total number of bls keys that signed the block.
+	TotalKeysSigned uint
+	// Mask is the bitmap Mask for the block.
+	Mask      *bls.Mask
+	BlockHash common.Hash
+}
+
+// GetDetailedBlockSignerInfo fetches the block signer information for any non-genesis block
+func (hmy *Harmony) GetDetailedBlockSignerInfo(
+	ctx context.Context, blk *types.Block,
+) (*DetailedBlockSignerInfo, error) {
+	slotList, mask, err := hmy.GetBlockSigners(
```

Comment:
will committee change per block, or just change per epoch?  will you consider to cache the data?

Reply:
It is cached...

The committee is fetched from shard state, which is cached.

https://github.com/harmony-one/harmony/blob/6ce8aeac3d448d0bc13e5f1286ce97856de42f1f/core/blockchain.go#L1800-L1816

https://github.com/harmony-one/harmony/blob/6ce8aeac3d448d0bc13e5f1286ce97856de42f1f/hmy/staking.go#L178-L190

https://github.com/harmony-one/harmony/blob/6ce8aeac3d448d0bc13e5f1286ce97856de42f1f/hmy/blockchain.go#L36-L53